pybio/spec/source.py

```diff
@@ -0,0 +1,273 @@
+            else:
+                # <module name>.<sub module name>.<object name>
+                last_dot_idx = uri.path.rfind(".")
+                self.module_name = uri.path[:last_dot_idx]
+                self.object_name = uri.path[last_dot_idx + 1 :]
+
+                if not re.match("[a-zA-Z_][a-zA-Z0-9_]*", self.object_name):
+                    raise ValueError(f"Invalid object name {self.object_name} extracted from source {source}")
+
+        elif uri.netloc == "github.com":
+            orga, repo_name, blob, commit_id, *in_repo_path = uri.path.strip("/").split("/")
+            in_repo_path = "/".join(in_repo_path)
+            cached_repo_path = self.cache_path / orga / repo_name
+            self.local_file_path = cached_repo_path / in_repo_path
+            cached_repo_path = cached_repo_path.resolve().as_posix()
+
+            self.download_commands = [
+                ["git", "clone", f"{uri.scheme}://{uri.netloc}/{orga}/{repo_name}.git", cached_repo_path],
+                # -C <working_dir> available in git 1.8.5+
+                # https://github.com/git/git/blob/5fd09df3937f54c5cfda4f1087f5d99433cce527/Documentation/RelNotes/1.8.5.txt#L115-L116
+                ["git", "-C", cached_repo_path, "checkout", "--force", commit_id]
+            ]
+
+        elif require_git:
+            raise ValueError(f"Invalid source: {source}")
+        elif uri.scheme in ["https", "http"]:
+            file_name = uri.path.strip("/")
+            self.local_file_path = self.cache_path / file_name
+        else:
+            raise ValueError(f"Unknown uri scheme {uri.scheme}")
+
+        self.uri = uri
+
+    @staticmethod
+    def resolve_doi(doi: str):
+        doi = doi.strip("/")
+
+        url = "https://doi.org/api/handles/" + doi
+        r = json.loads(requests.get(url).text)
+        response_code = r["responseCode"]
+        if response_code != 1:
+            raise RuntimeError(f"Could not resolve doi {doi} (responseCode={response_code})")
+
+        val = min(r["values"], key=lambda v: v["index"])
+
+        assert val["type"] == "URL"  # todo: handle other types
+        assert val["data"]["format"] == "string"
+        return urlparse(val["data"]["value"])
+
+    def download(self) -> None:
+        """download and cache source from doi or url"""
+        if self.local_file_path is None:
+            return
+
+        if self.local_file_path.exists():
+            # todo: check hash
+            return  # or not
+
+        assert self.cache_path in self.local_file_path.parents, "only download files to cache folder!"
+
+        uri = self.uri
+        if self.download_commands:
+            for command in self.download_commands:
+                call(command)
+
+            assert self.local_file_path.exists(), self.local_file_path
+            if self.object_name and self.object_name[0] == "L" and self.object_name[1:].isdigit():
+                # get object name from line number
```

Comment:
Why do we care about line number? Is that part of the spec?

Reply:
Is this to allow reading parts of modules hosted on github?

Reply:
yes, the fragment of a github url may contain the line number of the definition of that object. I thought this would be a neat way of defining importable objects... Having said that, this is deprecated, as we only get at yaml file here (and importable objects should always be specified as dependencies).

Reply:
ah, this is in source.py... this whole file is deprecated... I'll delete the old files..